### PR TITLE
Enable code coverage for ant, maven and gradle

### PR DESCRIPTION
**Pull request by AshwiniChalla**
Testing : unit tests, Manual testing. Ran all e2e tests against coreclr agent .

### CODE REVIEW

File: Tests/L0/ANT/_suite.ts [hunk 1]
```diff
@@ -262,5 +262,118 @@ describe('ANT Suite', function() {
...
+    
+    it('Ant calls enable code coverage and publish code coverage when Cobertura is selected.', (done) => {
+		setResponseFile('antCodeCoverage.json');
+		
+		var tr = new trm.TaskRunner('ANT');
+		tr.setInput('antBuildFile', '/build/build.xml'); // Make that checkPath returns true for this filename in the response file
+		tr.setInput('javaHomeSelection', 'JDKVersion');
+		tr.setInput('jdkVersion', 'default');
+		tr.setInput('testResultsFiles', '**/TEST-*.xml');
+        tr.setInput('codeCoverageTool', 'Cobertura');
```
Comment (nigurr) [inline Tests/L0/ANT/_suite.ts]: indention

File: Tests/L0/Gradle/_suite.ts [hunk 2]
```diff
@@ -6,416 +6,536 @@ import trm = require('../../lib/taskRunner');
...
+                assert(tr.resultWasSet, 'task should have set a result');
+                assert(tr.stderr.length == 0, 'should not have written to stderr');
+                assert(tr.succeeded, 'task should have succeeded');
+                done();
+            })
+            .fail((err) => {
+                done(err);
+            });
+    })
+
```
Comment (nigurr) [inline Tests/L0/Gradle/_suite.ts]: multi module test cases?

File: Tests/L0/Gradle/gradleGood.json [hunk 3]
```diff
@@ -4,6 +4,21 @@
             "code": 0,
             "stdout": "Sample gradle output"
         },
+        "gradlew properties": {
+            "code": 0,
+            "stdout": "Something subprojects",
```
Comment (nigurr) [inline Tests/L0/Gradle/gradleGood.json]: it should be "subprojects []" if it's single module project. this mock response is imp to test the code

File: Tests/L0/Gradle/_suite.ts (hunk 2 above)
Comment (AshwiniChalla) [inline Tests/L0/Gradle/_suite.ts]: added

File: Tasks/ANT/anttask.ts [hunk 4]
```diff
@@ -1,15 +1,18 @@
 /// <reference path="../../definitions/vsts-task-lib.d.ts" />
 
 import tl = require('vsts-task-lib/task');
```
Comment (nigurr) [inline Tasks/ANT/anttask.ts]: Copyright info

File: Tasks/ANT/anttask.ts [hunk 5]
```diff
@@ -1,15 +1,18 @@
...
+import path = require('path');
+import fs = require('fs');
 
 var anttool = tl.which('ant', true);
 
 var antv = tl.createToolRunner(anttool);
```
Comment (nigurr) [inline Tasks/ANT/anttask.ts]: Extra empty lines

**Comment by nigurr** [comment]
>     specifiedJavaHome = jdkUserInputPath;

unnecessary var

---
Refers to: Tasks/ANT/anttask.ts:59 in 98457c8. [](commit_id = 98457c84bc63b98ea5ad9026652a773163946af0, deletion_comment = False)

File: Tasks/ANT/anttask.ts [hunk 6]
```diff
@@ -61,44 +64,162 @@ if (specifiedJavaHome) {
...
-      var matchingTestResultsFiles = [testResultsFiles];
+}
+
+function enableCodeCoverage() {
+    var classFilter = tl.getInput('classFilter');
+    var classFilesDirectories = tl.getInput('classFilesDirectories', true);
+    var sourceDirectories = tl.getInput('srcDirectories');
+    // appending with small guid to keep it unique. Avoiding full guid to ensure no long path issues.
+    var reportDirectoryName = "CCReport43F6D5EF";
+    reportDirectory = path.join(buildRootPath, reportDirectoryName);
```
Comment (nigurr) [inline Tasks/ANT/anttask.ts]: hardcoded directory names?

File: Tasks/ANT/anttask.ts [hunk 7]
```diff
@@ -61,44 +64,162 @@ if (specifiedJavaHome) {
...
+    // clean any previous reports.
+    if (isFileExists(coberturaCCFile)) {
+        tl.rmRF(coberturaCCFile);
     }
 
-    if(!matchingTestResultsFiles) {
-      tl.warning('No test result files matching ' + testResultsFiles + ' were found, so publishing JUnit test results is being skipped.');  
-      return 0;
+    if (isDirectoryExists(reportDirectory)) {
+        tl.rmRF(reportDirectory);
```
Comment (nigurr) [inline Tasks/ANT/anttask.ts]: Do we need to check these? Remove directly

File: Tasks/ANT/anttask.ts [hunk 8]
```diff
@@ -61,44 +64,162 @@ if (specifiedJavaHome) {
...
-})
+function publishCodeCoverage(codeCoverageOpted: boolean) {
+    if (codeCoverageOpted) {
+        tl.debug("Collecting code coverage reports");
+        var antRunner = tl.createToolRunner(anttool);
+        if (isFileExists(reportBuildFile)) {
+            antRunner.arg('-buildfile');
+            antRunner.pathArg(reportBuildFile);
+            antRunner.arg(ccReportTask);
+        }
```
Comment (nigurr) [inline Tasks/ANT/anttask.ts]: refactor this code. based on condition you are just changing one argument.

File: Tasks/ANT/anttask.ts (hunk 6 above)
Comment (AshwiniChalla) [inline Tasks/ANT/anttask.ts]: explained offline

File: Tasks/ANT/anttask.ts [hunk 9]
```diff
@@ -1,15 +1,18 @@
 /// <reference path="../../definitions/vsts-task-lib.d.ts" />
 
 import tl = require('vsts-task-lib/task');
+import path = require('path');
```
Comment (nigurr) [inline Tasks/ANT/anttask.ts]: Get all user inputs in the top and assign to variables. Will keep code clean n organized.

Comment (nigurr) [inline Tasks/ANT/anttask.ts]: make sure we have unified "File Path separator" or we should support both *nix/Windows file separators in nodejs

**Comment by nigurr** [comment]
> import tl = require('vsts-task-lib/task');

Copyright

---
Refers to: Tasks/Gradle/gradletask.ts:3 in 98457c8. [](commit_id = 98457c84bc63b98ea5ad9026652a773163946af0, deletion_comment = False)

**Comment by nigurr** [comment]
>         "windows"

I guess We might need to remove this

---
Refers to: Tasks/Gradle/task.loc.json:193 in 98457c8. [](commit_id = 98457c84bc63b98ea5ad9026652a773163946af0, deletion_comment = False)

**Comment by nigurr** [comment]
> var testResultsFiles: string = tl.getInput('testResultsFiles', true);

This is clean. Get all inputs in the top and declare any global variables here

---
Refers to: Tasks/Maven/maventask.ts:12 in 98457c8. [](commit_id = 98457c84bc63b98ea5ad9026652a773163946af0, deletion_comment = False)

File: Tasks/ANT/anttask.ts (hunk 4 above)
Comment (AshwiniChalla) [inline Tasks/ANT/anttask.ts]: not needed

File: Tests/L0/Gradle/gradleCCSingleModule.json [hunk 10]
```diff
@@ -0,0 +1,52 @@
+{
+    "exec": {
+        "gradlew build": {
+            "code": 0,
+            "stdout": "Sample gradle output"
+        },
+        "gradlew properties": {
+            "code": 0,
+            "stdout": "Something subprojects",
```
Comment (nigurr) [inline Tests/L0/Gradle/gradleCCSingleModule.json]: "subprojects []

File: Tasks/ANT/anttask.ts (hunk 9 above)
Comment (AshwiniChalla) [inline Tasks/ANT/anttask.ts]: some user inputs ex: classFilesDirectory will be available only when code coverage is selected.

File: Tasks/ANT/ant.ps1 [hunk 11]
```diff
@@ -141,7 +139,7 @@ if($isCoverageEnabled)
 	# Enable code coverage in build file
 	if ($codeCoverageTool -eq "Cobertura")
 	{
-		$coberturaCCFile = Join-Path $buildRootPath "cobertura.cer"
+		$coberturaCCFile = Join-Path $buildRootPath "cobertura.ser"
```
Comment (acesiddhu) [inline Tasks/ANT/ant.ps1]: is this a issue in production also?

File: Tasks/ANT/anttask.ts (hunk 9 above)
Comment (AshwiniChalla) [inline Tasks/ANT/anttask.ts]: We have used path.join always so file path seperator doesnt come into picture here. Will change the task input help to always give linux style seperator.

File: Tasks/ANT/ant.ps1 (hunk 11 above)
Comment (AshwiniChalla) [inline Tasks/ANT/ant.ps1]: Ye. but this is not P0 as this is just cleanup.

File: Tasks/ANT/anttask.ts (hunk 7 above)
Comment (AshwiniChalla) [inline Tasks/ANT/anttask.ts]: done

File: Tasks/ANT/anttask.ts (hunk 8 above)
Comment (AshwiniChalla) [inline Tasks/ANT/anttask.ts]: done

File: Tasks/Maven/maventask.ts [hunk 12]
```diff
@@ -74,6 +74,21 @@ if (specifiedJavaHome) {
...
+var ccTool = tl.getInput('codeCoverageTool');
+var isCodeCoverageOpted = (typeof ccTool != "undefined" && ccTool && ccTool.toLowerCase() != 'none');
+
+if (isCodeCoverageOpted) {
+    var summaryFile = null;
+    var reportDirectory = null;
+    var reportPOMFile = null;
+    var execFileJacoco = null;
+    var ccReportTask = null;
+    enableCodeCoverage();
```
Comment (DavidStaheli) [inline Tasks/Maven/maventask.ts]: Can you add type information to the above vars?
`var summaryFile: string = null;`

File: Tasks/Maven/maventask.ts [hunk 13]
```diff
@@ -170,11 +186,121 @@ function publishJUnitTestResults(testResultsFiles: string) {
...
+function enableCodeCoverage() {
+    var classFilter = tl.getInput('classFilter');
+    var classFilesDirectories = tl.getInput('classFilesDirectories');
+    var sourceDirectories = tl.getInput('srcDirectories');
+    var buildRootPath = path.dirname(mavenPOMFile);
+    // appending with small guid to keep it unique. Avoiding full guid to ensure no long path issues.
+    var reportPOMFileName = "CCReportPomA4D283EG.xml";
+    reportPOMFile = path.join(buildRootPath, reportPOMFileName);
+    var targetDirectory = path.join(buildRootPath, "target");
+    ccReportTask = "jacoco:report";
```
Comment (DavidStaheli) [inline Tasks/Maven/maventask.ts]: Can you add type information to the above vars?
`var classFilter: string = tl.getInput('classFilter');`

File: Tasks/Gradle/gradletask.ts [hunk 14]
```diff
@@ -28,68 +28,180 @@ var javaHomeSelection = tl.getInput('javaHomeSelection', true);
...
+            var ccPublisher = new tl.CodeCoveragePublisher();
+            ccPublisher.publish(ccTool, summaryFile, reportDirectory, "");
+        }
+        else {
+            tl.warning("No code coverage results found to be published. This could occur if there were no tests executed or there was a build failure. Check the gradle output for details.");
+        }
+    }
+}
+
+function isFileExists(path: string) {
```
Comment (DavidStaheli) [inline Tasks/Gradle/gradletask.ts]: For clarity, this method should be named `pathExistsAsFile()`.

File: Tasks/Gradle/gradletask.ts [hunk 15]
```diff
@@ -28,68 +28,180 @@ var javaHomeSelection = tl.getInput('javaHomeSelection', true);
...
+    try {
+        return tl.stats(path).isFile();
+    }
+    catch (error) {
+        return false;
+    }
+}
+
+function isDirectoryExists(path: string) {
```
Comment (DavidStaheli) [inline Tasks/Gradle/gradletask.ts]: For clarity, this method should be named `pathExistsAsDirectory()`.

File: Tasks/Maven/maventask.ts [hunk 16]
```diff
@@ -170,11 +186,121 @@ function publishJUnitTestResults(testResultsFiles: string) {
...
+        tl.debug("Publishing code coverage results to TFS");
+        var ccPublisher = new tl.CodeCoveragePublisher();
+        ccPublisher.publish(ccTool, summaryFile, reportDirectory, "");
+    }
+    else {
+        tl.warning("No code coverage found to publish. There might be a build failure resulting in no code coverage or there might be no tests.");
+    }
+}
+
+function isFileExists(path: string) {
```
Comment (DavidStaheli) [inline Tasks/Maven/maventask.ts]: For clarity, this method should be named `pathExistsAsFile()`.

File: Tasks/Maven/maventask.ts [hunk 17]
```diff
@@ -170,11 +186,121 @@ function publishJUnitTestResults(testResultsFiles: string) {
...
+    try {
+        return tl.stats(path).isFile();
+    }
+    catch (error) {
+        return false;
+    }
+}
+
+function isDirectoryExists(path: string) {
```
Comment (DavidStaheli) [inline Tasks/Maven/maventask.ts]: For clarity, this method should be named `pathExistsAsDirectory()`.

File: Tasks/ANT/anttask.ts [hunk 18]
```diff
@@ -61,44 +63,151 @@ if (specifiedJavaHome) {
...
+        }).fail(function(err) {
+            tl.warning("No code coverage results found to be published. This could occur if there were no tests executed or there was a build failure. Check the ant output for details.");
+        });
     }
+}
 
-    var tp = new tl.TestPublisher("JUnit");
-    tp.publish(matchingTestResultsFiles, true, "", "","", true);
-  } 
+function isFileExists(path: string) {
```
Comment (DavidStaheli) [inline Tasks/ANT/anttask.ts]: For clarity, this method should be named `pathExistsAsFile()`.

File: Tasks/ANT/anttask.ts [hunk 19]
```diff
@@ -61,44 +63,151 @@ if (specifiedJavaHome) {
...
-  publishTestResults(publishJUnitResults, testResultsFiles);
-  tl.exit(code);
-})
-.fail(function(err) {
-  publishTestResults(publishJUnitResults, testResultsFiles);
-  console.error(err.message);
-  tl.debug('taskRunner fail');
-  tl.exit(1);
-})
+function isDirectoryExists(path: string) {
```
Comment (DavidStaheli) [inline Tasks/ANT/anttask.ts]: For clarity, this method should be named `pathExistsAsDirectory()`.

File: Tasks/Maven/maventask.ts [hunk 20]
```diff
@@ -170,11 +186,121 @@ function publishJUnitTestResults(testResultsFiles: string) {
...
+                tl.warning("No code coverage found to publish. There might be a build failure resulting in no code coverage or there might be no tests.");
+            });
+        }
+        else if (ccTool.toLowerCase() == "cobertura") {
+            publishCCToTfs();
+        }
+    }
+}
+
+function publishCCToTfs() {
```
Comment (DavidStaheli) [inline Tasks/Maven/maventask.ts]: It is confusing to have 2 functions named so similarly: publishCodeCoverage() and publishCCToTfs().  What if you did something like this?
```

function publishCodeCoverage(isCodeCoverageOpted: boolean) {
    if (isCodeCoverageOpted) {
        tl.debug("Collecting code coverage reports");
        if (********collectCodeCoverageReports()********) {
            if (isFileExists(summaryFile)) {
                tl.debug("Summary file = " + summaryFile);
                tl.debug("Report directory = " + reportDirectory);
                tl.debug("Publishing code coverage results to TFS");
                var ccPublisher = new tl.CodeCoveragePublisher();
                ccPublisher.publish(ccTool, summaryFile, reportDirectory, "");
            }
            else {
                tl.warning("No code coverage found to publish. There might be a build failure resulting in no code coverage or there might be no tests.");
            }
        }
    }
}
```

**Comment by DavidStaheli** [comment]
Thank you for adding the L0 tests!

File: Tasks/Maven/maventask.ts (hunk 20 above)
Comment (AshwiniChalla) [inline Tasks/Maven/maventask.ts]: This is done to refactor code.

**Pull request merged**